cmd/neofs-node/meta.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/nspcc-dev/neofs-node/pkg/core/container"
+	"github.com/nspcc-dev/neofs-node/pkg/core/netmap"
+	cntClient "github.com/nspcc-dev/neofs-node/pkg/morph/client/container"
+	"github.com/nspcc-dev/neofs-node/pkg/services/meta"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	netmapsdk "github.com/nspcc-dev/neofs-sdk-go/netmap"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+func initMeta(c *cfg) {
+	if c.cfgMorph.client == nil {
+		initMorphComponents(c)
+	}
+
+	c.cfgMeta.cLister = &containerListener{
+		key:        c.binPublicKey,
+		cnrClient:  c.basics.cCli,
+		containers: c.cfgObject.cnrSource,
+		network:    c.basics.netMapSource,
+	}
+
+	m, err := meta.New(c.log.With(zap.String("service", "meta data")),
+		c.cfgMeta.cLister,
+		c.applicationConfiguration.fsChain.dialTimeout,
+		c.applicationConfiguration.fsChain.endpoints,
+		c.basics.containerSH,
+		c.basics.netmapSH,
+		c.applicationConfiguration.metadata.path)
+	fatalOnErr(err)
+
+	c.workers = append(c.workers, newWorkerFromFunc(func(ctx context.Context) {
+		err = m.Run(ctx)
+		if err != nil {
+			c.internalErr <- fmt.Errorf("meta data service error: %w", err)
+		}
+	}))
+}
+
+type containerListener struct {
+	key []byte
+
+	cnrClient  *cntClient.Client
+	containers container.Source
+	network    netmap.Source
+
+	m          sync.RWMutex
+	prevCnrs   []cid.ID
+	prevNetMap *netmapsdk.NetMap
+	prevRes    map[cid.ID]struct{}
+}
+
+func (c *containerListener) List() (map[cid.ID]struct{}, error) {
+	actualContainers, err := c.cnrClient.List(nil)
```

Comment:
have u considered https://github.com/nspcc-dev/neofs-contract/blob/27ef3cc95fd1b601845effe6b5b6f36b30744836/contracts/container/contract.go#L578 for this?

Reply:
yes, but i think it should be fixed once for all callers: https://github.com/nspcc-dev/neofs-node/issues/3045